spec/firebase_app_distribution_api_client_spec.rb

```diff
@@ -128,7 +127,92 @@
   end
 
   describe '#upload' do
-    # Empty for now
+    let(:upload_status_response_success) do
+      UploadStatusResponse.new(
+        { status: "SUCCESS",
+          release: { id: "release_id" } }
+      )
+    end
+    let(:upload_status_response_error) do
+      UploadStatusResponse.new(
+        { status: "ERROR",
+          release: {} }
+      )
+    end
+
+    before(:each) do
+      # Stub out polling interval for quick specs
+      stub_const("Fastlane::Client::FirebaseAppDistributionApiClient::POLLING_INTERVAL_SECONDS", 0)
+
+      # Expect a call to get_upload_token every time
+      expect(api_client).to receive(:get_upload_token)
+        .with("app_id", fake_binary_path)
+        .and_return("upload_token").ordered
+    end
+
+    it 'skips the upload step if the binary has already been uploaded' do
+      # upload should not attempt to upload the binary at all
+      expect(api_client).not_to(receive(:upload_binary))
+      expect(api_client).to receive(:get_upload_status)
+        .with("app_id", "upload_token")
+        .and_return(upload_status_response_success)
+
+      release_id = api_client.upload("app_id", fake_binary_path)
+      expect(release_id).to eq("release_id")
+    end
+
+    it 'uploads the app binary then returns the release_id' do
+      # return an error then a success after being uploaded
+      expect(api_client).to receive(:get_upload_status)
+        .with("app_id", "upload_token")
+        .and_return(upload_status_response_error, upload_status_response_success).ordered
+
+      # upload_binary should only be called once
+      expect(api_client).to receive(:upload_binary)
+        .with("app_id", fake_binary_path).ordered
+
+      release_id = api_client.upload("app_id", fake_binary_path)
+      expect(release_id).to eq("release_id")
+    end
+
+    it 'attempts to upload MAX_POLLING_RETRIES times' do
```

Comment:
I don't think we should be attempting to upload the binary `MAX_POLLING_RETRIES` times. After uploading the binary, we should be polling for the upload status `MAX_POLLING_RETRIES` times.

Reply:
That's right, I'll change the current test and method description